File: src/koffi/test/async.js
#!/usr/bin/env node
// SPDX-License-Identifier: MIT
// SPDX-FileCopyrightText: 2025 Niels Martignène <niels.martignene@protonmail.com>

const koffi = require('../../koffi');
const assert = require('assert');
const path = require('path');
const { cnoke } = require('./package.json');

const PackedBFG = koffi.pack('PackedBFG', {
    a: 'int8_t',
    b: 'int64_t',
    c: 'char',
    d: 'str',
    e: 'short',
    inner: koffi.pack({
        f: 'float',
        g: 'double'
    })
});

const CharCallback = koffi.proto('int CharCallback(int idx, char c)');
const BinaryIntFunc = koffi.proto('int BinaryIntFunc(int a, int b)');

main();

async function main() {
    try {
        await test();
        console.log('Success!');
    } catch (err) {
        console.error(err);
        process.exit(1);
    }
}

async function test() {
    const lib_filename = path.join(__dirname, cnoke.output, 'async' + koffi.extension);
    const lib = koffi.load(lib_filename);

    const ConcatenateToInt1 = lib.func('ConcatenateToInt1', 'int64_t', Array(12).fill('int8_t'));
    const MakePackedBFG = lib.func('PackedBFG __fastcall MakePackedBFG(int x, double y, _Out_ PackedBFG *p, const char *str)');
    const CallMeChar = lib.func('int CallMeChar(CharCallback *func)');
    const GetBinaryIntFunction = lib.func('BinaryIntFunc *GetBinaryIntFunction(const char *name)');

    let promises = [];

    // Issue several async calls
    for (let i = 0; i < 32; i++) {
        let p = new Promise((resolve, reject) => {
            try {
                ConcatenateToInt1.async(5, 6, 1, 2, 3, 9, 4, 4, 0, 6, 8, 7, (err, res) => {
                    assert.equal(err, null);
                    assert.equal(res, 561239440687n);
                });
                resolve();
            } catch (err) {
                reject(err);
            }
        });
        promises.push(p);
    }

    // Make a few synchronous calls while the asynchronous ones are still in the air
    assert.equal(ConcatenateToInt1(5, 6, 1, 2, 3, 9, 4, 4, 0, 6, 8, 7), 561239440687n);
    assert.equal(ConcatenateToInt1(5, 6, 1, 2, 3, 9, 4, 4, 0, 6, 8, 7), 561239440687n);
    assert.equal(ConcatenateToInt1(5, 6, 1, 2, 3, 9, 4, 4, 0, 6, 8, 7), 561239440687n);

    // Async complex call
    {
        let out = {};
        let p = new Promise((resolve, reject) => {
            try {
                MakePackedBFG.async(2, 7, out, '__Hello123456789++++foobarFOOBAR!__', (err, res) => {
                    assert.deepEqual(res, { a: 2, b: 4, c: -25, d: 'X/__Hello123456789++++foobarFOOBAR!__/X', e: 54, inner: { f: 14, g: 5 } });
                    assert.deepEqual(out, res);
                });
                resolve();
            } catch (err) {
                reject(err);
            }
        });
        promises.push(p);
    }

    // Call function pointers
    {
        test_binary('add', 4, 5, 9);
        test_binary('substract', 4, 5, -1);
        test_binary('multiply', 3, 8, 24);
        test_binary('divide', 100, 2, 50);

        function test_binary(type, a, b, expected) {
            let p = new Promise((resolve, reject) => {
                try {
                    let ptr = GetBinaryIntFunction(type);
                    let func = koffi.decode(ptr, BinaryIntFunc);

                    func.async(a, b, (err, res) => {
                        assert.equal(res, expected);
                    });

                    resolve();
                } catch (err) {
                    reject(err);
                }
            });
            promises.push(p);
        }
    }

    await Promise.all(promises);
}

File: src/koffi/test/async.c
// SPDX-License-Identifier: MIT
// SPDX-FileCopyrightText: 2025 Niels Martignène <niels.martignene@protonmail.com>

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <inttypes.h>

#if defined(_WIN32)
    #define EXPORT __declspec(dllexport)
#else
    #define EXPORT __attribute__((visibility("default")))
#endif
#if defined(_M_IX86) || defined(__i386__)
    #if defined(_MSC_VER)
        #define FASTCALL __fastcall
        #define STDCALL __stdcall
    #else
        #define FASTCALL __attribute__((fastcall))
        #define STDCALL __attribute__((stdcall))
    #endif
#else
    #define FASTCALL
    #define STDCALL
#endif

#pragma pack(push, 1)
typedef struct PackedBFG {
    int8_t a;
    int64_t b;
    signed char c;
    const char *d;
    short e;
    struct {
        float f;
        double g;
    } inner;
} PackedBFG;
#pragma pack(pop)

typedef int CharCallback(int idx, char c);
typedef int BinaryIntFunc(int a, int b);

EXPORT int64_t ConcatenateToInt1(int8_t a, int8_t b, int8_t c, int8_t d, int8_t e, int8_t f,
                                 int8_t g, int8_t h, int8_t i, int8_t j, int8_t k, int8_t l)
{
    int64_t ret = 100000000000ull * a + 10000000000ull * b + 1000000000ull * c +
                  100000000ull * d + 10000000ull * e + 1000000ull * f +
                  100000ull * g + 10000ull * h + 1000ull * i +
                  100ull * j + 10ull * k + 1ull * l;
    return ret;
}

EXPORT PackedBFG FASTCALL MakePackedBFG(int x, double y, PackedBFG *p, const char *str)
{
    PackedBFG bfg;

    static char buf[64];
    snprintf(buf, sizeof(buf), "X/%s/X", str);

    bfg.a = x;
    bfg.b = x * 2;
    bfg.c = x - 27;
    bfg.d = buf;
    bfg.e = x * 27;
    bfg.inner.f = (float)y * x;
    bfg.inner.g = (double)y - x;
    *p = bfg;

    return bfg;
}

EXPORT int CallMeChar(CharCallback *func)
{
    int ret = 0;

    ret += func(0, 'a');
    ret += func(1, 'b');

    return ret;
}

static int AddInt(int a, int b) { return a + b; }
static int SubstractInt(int a, int b) { return a - b; }
static int MultiplyInt(int a, int b) { return a * b; }
static int DivideInt(int a, int b) { return a / b; }

EXPORT BinaryIntFunc *GetBinaryIntFunction(const char *type)
{
    if (!strcmp(type, "add")) {
        return AddInt;
    } else if (!strcmp(type, "substract")) {
        return SubstractInt;
    } else if (!strcmp(type, "multiply")) {
        return MultiplyInt;
    } else if (!strcmp(type, "divide")) {
        return DivideInt;
    } else {
        return NULL;
    }
}
